### PR TITLE
Feature/381 upgrade cluster

### LINKED ISSUE #381: Add Route to get upgradable versions
We want a route, that responses the possible upgradable versions and we need a route, which performs the upgrade to a specific version.

### PR DESCRIPTION
**Pull request by guusvw**
Close #381

### CODE REVIEW

File: api/cmd/kubermatic-api/main.go [hunk 1]
```diff
@@ -27,6 +29,8 @@ var (
 	address          = flag.String("address", ":8080", "The address to listen on")
 	masterKubeconfig = flag.String("master-kubeconfig", "", "When set it will overwrite the usage of the InClusterConfig")
 	tokenIssuer      = flag.String("token-issuer", "", "URL of the OpenID token issuer. Example: http://auth.int.kubermatic.io")
+	versionsFile     = flag.String("versions", "versions.yaml", "The versions.yaml file path")
+	updatesFile      = flag.String("updates", "updates.yaml", "The updates.yaml file path")
```
Comment (mrIncompetent) [inline api/cmd/kubermatic-api/main.go:35]: Deployment file also needs to be updated!

File: api/handler/upgrade.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+			return nil, err
+		}
+
+		_, ok = versions[req.to]
+		if !ok {
+			return nil, NewUnknownVersion(req.to)
+		}
+
+		c.Spec.MasterVersion = req.to
+		c.Status.Phase = api.UpdatingMasterClusterStatusPhase
```
Comment (mrIncompetent) [inline api/handler/upgrade.go]: After setting the phase, you also need to persist. 
Also: This should probably be moved into the kubernetes provider

**Review by mrIncompetent — CHANGES_REQUESTED** [review]

File: api/cmd/kubermatic-api/main.go (hunk 1 above)
Comment (guusvw) [reply api/cmd/kubermatic-api/main.go:35]: Done in https://github.com/kubermatic/kubermatic/pull/382/commits/2148d257d2233ba458a47bab3cb15862219750a2

File: api/provider/kubernetes/kubernetes.go [hunk 3]
```diff
@@ -393,3 +393,21 @@ func (p *kubernetesProvider) DeleteCluster(user provider.User, cluster string) e
 
 	return p.kuberntesClient.Namespaces().Delete(NamespaceName(cluster), &metav1.DeleteOptions{})
 }
+
+func (p *kubernetesProvider) UpgradeCluster(user provider.User, cluster, version string) error {
```
Comment (guusvw) [inline api/provider/kubernetes/kubernetes.go:397]: @mrIncompetent is this the way to go?

File: api/handler/upgrade.go (hunk 2 above)
Comment (guusvw) [reply api/handler/upgrade.go]: Moved: https://github.com/kubermatic/kubermatic/pull/382/commits/9f9eb31e3d65d6f6f5315a4b7acf43850dfe3652#diff-1701e2bab95c63ac3a77bc5844994042L108

https://github.com/kubermatic/kubermatic/pull/382/commits/9f9eb31e3d65d6f6f5315a4b7acf43850dfe3652#diff-db54f997140cb0ce93fd32e0dbf6826cR397

File: api/cmd/kubermatic-api/main.go [hunk 4]
```diff
@@ -71,7 +75,20 @@ func main() {
...
+	versions := make(map[string]*api.MasterVersion)
+	if *versionsFile != "" {
+		var err error
+		versions, err = version.LoadVersions(*updatesFile)
+		if err != nil {
+			glog.Fatal(fmt.Sprintf("failed to load updates yaml %q: %v", *updatesFile, err))
+		}
+
+		glog.Fatal(fmt.Sprintf("failed to load version yaml %q: %v", *versionsFile, err))
+	}
```
Comment (mrIncompetent) [inline api/cmd/kubermatic-api/main.go]: We should not continue if the versions-file or updates-file is not defined

File: api/cmd/kubermatic-api/main.go [hunk 5]
```diff
@@ -71,7 +75,20 @@ func main() {
 
 	// start server
 	ctx := context.Background()
-	r := handler.NewRouting(ctx, dcs, kps, cps, authenticator, masterTPRClient)
+
+	// load versions
+	versions := make(map[string]*api.MasterVersion)
+	if *versionsFile != "" {
+		var err error
+		versions, err = version.LoadVersions(*updatesFile)
```
Comment (mrIncompetent) [inline api/cmd/kubermatic-api/main.go]: updatesFile & versionsFile?
Why checking versionsFile to load updatesFile?

File: api/cmd/kubermatic-cluster-controller/main.go [hunk 6]
```diff
@@ -49,8 +49,15 @@ func main() {
 	}
 
 	// load versions
-	versions, err := version.LoadVersions(*versionsFile)
-	if err != nil {
+	versions := make(map[string]*api.MasterVersion)
+	if *versionsFile != "" {
+
+		var err error
+		versions, err = version.LoadVersions(*updatesFile)
```
Comment (mrIncompetent) [inline api/cmd/kubermatic-cluster-controller/main.go]: updatesFile & versionsFile?
Why checking versionsFile to load updatesFile?

File: api/cmd/kubermatic-cluster-controller/main.go [hunk 7]
```diff
@@ -49,8 +49,15 @@ func main() {
...
+		if err != nil {
+			glog.Fatal(fmt.Sprintf("failed to load updates yaml %q: %v", *updatesFile, err))
+		}
+
 		glog.Fatal(fmt.Sprintf("failed to load version yaml %q: %v", *versionsFile, err))
 	}
```
Comment (mrIncompetent) [inline api/cmd/kubermatic-cluster-controller/main.go]: We should not continue if the versions-file or updates-file is not defined

File: api/handler/upgrade.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+			}
+			return nil, err
+		}
+
+		current, err := version.NewVersion(c.Spec.MasterVersion)
+		if err != nil {
+			return nil, err
+		}
+
+		possibleUpdates := make(version.Collection, 0)
```
Comment (mrIncompetent) [inline api/handler/upgrade.go]: I think we should use the updates.yaml so we only show possible

File: api/provider/kubernetes/kubernetes.go (hunk 3 above)
Comment (mrIncompetent) [reply api/provider/kubernetes/kubernetes.go:397]: Yes!

**Review by mrIncompetent — CHANGES_REQUESTED** [review]

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**